### ISSUE TITLE
List of links (topology) should be in core and not in topology

### ISSUE THREAD

**Issue by ajoaoff** [kytos-ng/kytos#92]
Original issue opened by @ajoaoff at https://github.com/kytos/kytos/issues/1064.

The list of links must be accessible by all applications, like the list of switches and interfaces. Although the topology is discovered by the `kytos/topology` NApp, the results should be stored in the core. The NApp `kytos/of_core` deals with switch connections, but the data is stored in the core.

**Comment by viniarck** [comment]
Related issues:

https://github.com/kytos-ng/maintenance/issues/90

https://github.com/kytos-ng/mef_eline/issues/464

**Comment by viniarck** [comment]
Quick summary of major points to be moved and refactored on both `kytos` and `topology`:

- Move `_get_link_or_create(endpoint_a, endpoint_b)` to `kytos.core.controller.get_link_or_create(endpoint_a: Interface, endpoint_b: Interface)`
- Move `topology` `self.links` dict to  `Controller.links`
- Move `topology` `self._links_lock` to `Controller._links_lock`.
- Move `topology` `get_link_from_interface(interface)` to `kytos.core.controller.get_link_from_interface(interface: Interface) -> Optional[Link]`. This method should be thread safe using `Controller._links_lock`.
- Add a new method `get_link(link_id: str) -> Optional[Link]`
- Refactor all topology `self.links` usage to use `kytos.core.controller.links` instead, including preserving the shallow copies
- Refactor topology `self._links_lock` to use a shallow copy where it's iterating over links.
- Refactor topology `on_interruption_start` to use another local lock something like `with self._interruption_lock` each start event needs to be processed sequentially, and `on_interruption_end` to also use the same lock.
- Add a new method on `kytos.core.controller.remove_link(link_id: str) -> Optional[Link]`

All topology DB writes will still be managed by topology, only the `Link` instances refs should be moved to core just so certain memory reads becomes easier and avoid IO in many cases that it used to have in certain request calls.

`topology` endpoints will still stay the same, no compatibility broken except `self.links` which shouldn't accessed by other NApps, although `maintenance` is doing it https://github.com/kytos-ng/maintenance/issues/90, so that needs to be refactored too.

**Comment by Alopalao** [comment]
In addition to the summary, I think it will be good idea to add these:
- A `Lock` from `Link` to be called with `link.link_lock`. Previously `self._links_lock` was being used but since it is removed for single link operations we can called the `Link` `self.link_lock` instead
- Move these [lines](https://github.com/kytos-ng/topology/blob/1ddf8c8c82a242942e915302a90739b1f9bb1490/main.py#L1203-L1210) to `_get_link_or_create` and use `new_link.link_lock`.
One observation only:
- Interface status also modifies its link status. But any modification to an Interface status will create an event which we handle. So far there is no need for an interface lock.

**Comment by viniarck** [comment]
> In addition to the summary, I think it will be good idea to add these:
> 
> * A `Lock` from `Link` to be called with `link.link_lock`. Previously `self._links_lock` was being used but since it is removed for single link operations we can called the `Link` `self.link_lock` instead

Yes @Alopalao, good observation. Let's go for it, and skill keep `topology` acquiring this lock in the places where it was requiring a lock for the link. Where it was iterating though, let's keep the shallow copies.

> * Move these [lines](https://github.com/kytos-ng/topology/blob/1ddf8c8c82a242942e915302a90739b1f9bb1490/main.py#L1203-L1210) to `_get_link_or_create` and use `new_link.link_lock`.
> 

Good finding; +1.
